### PR TITLE
A post and 2 tutorials on single-file open geospatial standards

### PR DESCRIPTION
**Pull request by florisvdh**
I believe it is time to introduce these formats to scientists so that they learn and implement their usage.

### CODE REVIEW

**Comment by florisvdh** [comment]
To ease reviewing, [herewith](https://github.com/inbo/tutorials/files/3874140/geospatial_standards.zip) the compiled html versions.

File: content/articles/geospatial_standards/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+], 
+so the GeoPackage is a more than suitable replacement.
+- The GeoPackage can store one or _multiple_ **vector** layers (points, lines, polygons and related feature types).
+Beside vector data, it can also store **raster** data^[
+For larger rasters however, the GeoTIFF file format will need less disk space and is therefore recommended.
+]
+or extra standalone **tables**.
+- The GeoPackage standard is [maintained](https://www.opengeospatial.org/standards/geopackage) by the [Open Geospatial Consortium](https://www.opengeospatial.org/) (OGC), which stands out as a reference when it comes to open geospatial standards.
+
+## The GeoJSON file format
```
Comment (hansvancalster) [inline content/articles/geospatial_standards/index.Rmd]: For ArcGis users: this is the equivalent of a .shp file (and it's associated files)

File: content/articles/geospatial_standards/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+- In order to keep it manageable (text file size, usage in versioning systems) it can be wise to use GeoJSON for more simple cases (points and rather simple lines and polygons), and use the binary GeoPackage format for larger (more complex) cases.
+
+## The GeoTIFF file format
+
+- [GeoTIFF](https://en.wikipedia.org/wiki/GeoTIFF) is the preferred single-file open standard for **raster** data.
+It adheres to the open [TIFF](https://en.wikipedia.org/wiki/TIFF) specification; hence it is a TIFF image file (`filename.tif`).
+It [uses](http://docs.opengeospatial.org/is/19-008r4/19-008r4.html#_geotiff_file_structure_and_geotiff_crs_and_models_principles_informative) a small set of reserved TIFF tags to store information about CRS, extent and resolution of the raster.
+- A GeoTIFF file can contain _one_ or _multiple_ rasters with the same CRS, extent and resolution.
+- The GeoTIFF standard is [maintained](https://www.opengeospatial.org/standards/geotiff) by the [Open Geospatial Consortium](https://www.opengeospatial.org/) (OGC), which stands out as a reference when it comes to open geospatial standards.
```
Comment (hansvancalster) [inline content/articles/geospatial_standards/index.Rmd]: Too much information?

File: content/tutorials/spatial_standards_raster/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+While only a few functions are used, you can use the previous hyperlinks to access the tutorials (vignettes) of these packages for more functions and information.
+
+```{r warning=FALSE, message=FALSE}
+options(stringsAsFactors = FALSE)
+library(raster)
+library(tidyverse)
+library(inborutils)
+```
+
+You find bit more background about 'why and what', regarding the open GeoTIFF standard, in [a separate post](../../articles/geospatial_standards/) on this website.
```
Comment (hansvancalster) [inline content/tutorials/spatial_standards_raster/index.Rmd]:
```suggestion
You will find a bit more background about 'why and what', regarding the open GeoTIFF standard, in [a separate post](../../articles/geospatial_standards/) on this website.
```

File: content/tutorials/spatial_standards_raster/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+In short, the GeoTIFF format is ideal for exchange, publication, interoperability & durability and to open science in general.
+
+## How to make and use GeoTIFF files (`*.tif`)
```
Comment (hansvancalster) [inline content/tutorials/spatial_standards_raster/index.Rmd]:
```suggestion
## How to make and use of GeoTIFF files (`*.tif`)
```

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+library(tidyverse)
+library(sf)
+library(inborutils)
+```
+
+You find a bit more background about 'why and what', regarding the below open standards, in [a separate post](../../articles/geospatial_standards/) on this website.
+
+In short, the GeoPackage and GeoJSON formats are ideal for exchange, publication, interoperability & durability and to open science in general.
+
+## How to make and use GeoPackages (`*.gpkg`)
```
Comment (hansvancalster) [inline content/tutorials/spatial_standards_vector/index.Rmd]:
```suggestion
## How to make and use of GeoPackages (`*.gpkg`)
```

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+### Making a GeoPackage from a geospatial `sf` object in R
```
Comment (hansvancalster) [inline content/tutorials/spatial_standards_vector/index.Rmd]: I suggest to move this section to the end.

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+Is that all?\
+_**YES :-)**_
+
+Really?\
+_**YES :-)**_
+
+Well, hmmm, if you really want to know a little bit more...
+
+A GeoPackage can contain many layers.
```
Comment (hansvancalster) [inline content/tutorials/spatial_standards_vector/index.Rmd]: Would love to see this demonstrated as well. Maybe even in combination with a raster?

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+### Reading a GeoPackage file
+
+Can it become more simple than this?
+
+```{r}
+sac_test <- st_read("sac.gpkg")
+```
+
+Ready you are!
```
Comment (hansvancalster) [inline content/tutorials/spatial_standards_vector/index.Rmd]:
```suggestion
Ready!
```

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+```{r echo=FALSE, comment=""}
+readLines("habitatstreams.geojson", n = 7) %>% 
+  cat(sep = "\n")
+```
+
+You can see it basically lists the feature attributes and the coordinates of the lines' vertices, with each feature starting on a new line.
+
+Compare the coordinates with those of the `sf` object `habitatstreams` above: the data have been reprojected on the fly to WGS84!
+
+Note: in order to be still manageable (text file size, usage in versioning systems) it seems wise to use GeoJSON for more simple cases -- **points and rather simple lines and polygons** -- and use the binary GeoPackage format for larger (more complex) cases.
```
Comment (hansvancalster) [inline content/tutorials/spatial_standards_vector/index.Rmd]: I think this recommendation and the above (geoJSON can be used in version control) should stand out more prominently. Maybe in a different section in the beginning of the document?

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+Just do this:
+
+```{r}
+habitatstreams_test <- st_read("habitatstreams.geojson")
+```
+
+Same story as for the GeoPackage: other geospatial software will (or should be) able to open the GeoJSON format as well, as it's an open and well established standard.
+
+From the message of `st_read()` you can see the CRS is WGS84 ([EPSG-code 4326](https://epsg.io/4326)) - this is always expected when reading a GeoJSON file.
```
Comment (hansvancalster) [inline content/tutorials/spatial_standards_vector/index.Rmd]: This is another difference between both formats that can be added to a separate section that compares both and recommends when to use which.

**Comment by florisvdh** [comment]
Thanks @hansvancalster for reviewing. Regarding putting GPKG after GeoJSON, I have some counterideas, so I'll await other people's ideas:

- GeoPackage is more universal to all usecases, and much propagated by the developers and writers in the R Spatial ecosystem as a replacement of shapefiles. For example, it led Roger Bivand to move his `rgrass7` package to use GPKG as an intermediate format, instead of shapefiles.
- GeoJSON, though universal and having a simple approach (text), brings its own limitations: it's not a database, it gets inflated with more complex data. And for the tutorial it needs a bit more intricate explanation than GPKG (which works out of the box) because its versions / CRS approach.

It's a good idea to show how to add multiple layers in GPKG, I'll add that.

Regarding rasters in a GPKG, it seems I only tried it with a GRTS master grid - and that means my footnote in the article has a mistake (I've messed up with RasterLite here, will fix it). I was coping [at that time](https://github.com/inbo/n2khab-preprocessing/blob/master/src/miscellaneous/GRTSmh.Rmd#L1163) with very large numbers (Int32 GRTS addresses), which is supported by GeoTIFF but not by GPKG. (And the RasterLite alternative got so inflated with larger rasters (file size) compared to GeoTIFF (really a lot) that that option was not good enough to me).

So yes, I'll give GPKG another try for rasters, with more simple data; I'm curious whether it is better at filesize than RasterLite (which is also an SQLite implementation).

Anyhow, for vector data GPKG was [much more efficient](https://github.com/inbo/n2khab-preprocessing/blob/146e115163996efef3429d156453a7ac2c79f330/src/miscellaneous/GRTSmh.Rmd#L1434-L1440) in filesize than SpatiaLite (another SQLite implementation). Both RasterLite and SpatialLite exist longer but clearly GPKG is now preferred and it is the only one which is embraced and maintained by the OGC consortium.

File: content/articles/geospatial_standards/index.Rmd [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+Using GDAL is the easiest way to conform to open standards.
+
+So, in R we use packages that use GDAL in the background, such as `rgdal`, `sp`, `sf`, `raster` and `stars`.
+
+## The GeoPackage file format
+
+- Its website is <https://www.geopackage.org>.
+- It is a standardized implementation of an SQLite database for geospatial data.
+Hence, a GeoPackage is a **binary** file (`filename.gpkg`).
+It shares this property with shapefiles, which however pose multiple limitations ^[
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: shapefiles is plural, so they 'pose' limitations

File: content/tutorials/spatial_standards_raster/index.Rmd (hunk 4 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_raster/index.Rmd]: it is not 'how to make use of' but 'make + use'

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+While only a few functions are used, you can use the previous hyperlinks to access the tutorials (vignettes) of these packages for more functions and information.
+
+```{r warning=FALSE, message=FALSE}
+options(stringsAsFactors = FALSE)
+library(tidyverse)
+library(sf)
+library(inborutils)
+```
+
+You find a bit more background about 'why and what', regarding the below open standards, in [a separate post](../../articles/geospatial_standards/) on this website.
```
Comment (florisvdh) [inline content/tutorials/spatial_standards_vector/index.Rmd]:
```suggestion
You will find a bit more background about 'why and what', regarding the below open standards, in [a separate post](../../articles/geospatial_standards/) on this website.
```

File: content/articles/geospatial_standards/index.Rmd [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+- [a lot](https://gdal.org/drivers/raster/index.html) of raster formats;
+- [a lot](https://gdal.org/drivers/vector/index.html) of vector formats.
+
+In other words, it is the preferred workhorse for reading and writing many geospatial file formats, used in the background by [a lot](https://gdal.org/software_using_gdal.html#software-using-gdal) of geospatial applications.
+Using GDAL is the easiest way to conform to open standards.
+
+So, in R we use packages that use GDAL in the background, such as `rgdal`, `sp`, `sf`, `raster` and `stars`.
+
+## The GeoPackage file format
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Sounds good to me. However, at an ArcGis [site](https://desktop.arcgis.com/en/arcmap/10.3/manage-data/administer-file-gdbs/file-geodatabases.htm) I read:

>A file geodatabase is a collection of files in a folder on disk that can store, query, and manage both spatial and nonspatial data.

So maybe not that similar, after all.

File: content/articles/geospatial_standards/index.Rmd [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+]
+or extra standalone **tables**.
+- The GeoPackage standard is [maintained](https://www.opengeospatial.org/standards/geopackage) by the [Open Geospatial Consortium](https://www.opengeospatial.org/) (OGC), which stands out as a reference when it comes to open geospatial standards.
+
+## The GeoJSON file format
+
+- One [GeoJSON](https://tools.ietf.org/html/rfc7946) file (`filename.geojson`) contains _one_ **vector** layer.
+[JSON](https://en.wikipedia.org/wiki/JSON) itself is a common and straightforward open data format.
+It is a **text** file readable both by humans and machines (see the [tutorial](../../tutorials/spatial_standards_vector/) for an example).
+GeoJSON adds the necessary specification to JSON for standardized storage of geographic feature data, but it is still a plain JSON text file.
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Understandable. On the other hand, that's why I kept this out of the tutorials and put it in an article. I believe it's good to provide limited background.

File: content/articles/geospatial_standards/index.Rmd (hunk 1 above)
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Rather, the GeoPackage is the equivalent of the shapefile. Binary, self-contained, exchangeable. That's also how it's connected in literature.

But, it's not a bad idea to more explicitly state this and address ArcGis users indeed.

GeoJSON is really something of its own (text).

File: content/articles/geospatial_standards/index.Rmd (hunk 2 above)
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: I don't believe so, it is good to be aware of what the data formats are. Rather, I may even have missed important aspects, perhaps. This is the article, not the tutorials.

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 7 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: Let's try.

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 9 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: Good idea, will try.

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 10 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: The recommendation to make a comparative section sounds good. Will try.

**Comment by florisvdh** [comment]
Long live git. It seems I did an experiment with rasters in GPKG (although obsoleted it because of the unsupported Int32 type). See [these lines](https://github.com/inbo/n2khab-preprocessing/blob/715c4271500881499ffe6744c001578234779288/src/miscellaneous/GRTSmh.Rmd#L1144-L1247). And the conclusion was: GPKG filesize can even be **smaller** than GeoTIFF :smiley: !

Some work is to be done, obviously. It may even be better to split off GPKG in a separate tutorial, combining vector & raster, or perhaps just add another GPKG section in the raster tutorial. Actually I prefer the latter, hence keeping vector & raster separated; mixing up both will complicate matters and make it more difficult to the reader.

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 7 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: This has been added in d5aa961

File: content/articles/geospatial_standards/index.Rmd (hunk 14 above)
[the remark this answers is not in the document]
Comment (hansvancalster) [reply content/articles/geospatial_standards/index.Rmd]: OK

File: content/articles/geospatial_standards/index.Rmd (hunk 1 above)
Comment (hansvancalster) [reply content/articles/geospatial_standards/index.Rmd]: I wouldn't call GeoPackage the equivalent of a shapefile, because it can contain multiple shape and raster files.

File: content/articles/geospatial_standards/index.Rmd (hunk 2 above)
Comment (hansvancalster) [reply content/articles/geospatial_standards/index.Rmd]: Agreed, definitely keep it in the article

File: content/articles/geospatial_standards/index.Rmd (hunk 1 above)
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: That is indeed something extra for GPKG, there you have a point. In that sense, I think they're all different things. Anyhow, you'll find a lot of places where GPKG is seen as a successor to the shapefile (e.g. [1](https://geocompr.robinlovelace.net/read-write.html#file-formats), [2](https://feed.terramonitor.com/shapefile-vs-geopackage-vs-geojson/), [3](https://www.gis-blog.com/geopackage-vs-shapefile/)), but [here](https://carto.com/blog/fgdb-gpkg/) is an interesting one where FGDB is also discussed, next to shapefile & GPKG.

Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: The following (from [this source](https://carto.com/blog/fgdb-gpkg/)) could be an important aspect where FGDB differs in terms of portability, with shapefile & GPKG:

> any FGDB data you receive will be a zip file that unzips to a .gdb directory

One can attach files to an email, but FGDB seems to require a folder setup as well.

File: content/tutorials/spatial_standards_raster/index.Rmd [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+arts %>% 
+  as.list %>% 
+  lapply(spplot)
+```
+
+So now what?\
+_**Let's write it!**_
+
+```{r}
+arts %>%
```
Comment (ThierryO) [inline content/tutorials/spatial_standards_raster/index.Rmd]: piping seems a bit overkill in these simple cases

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+```
+
+Note, `delete_dsn` was set as `TRUE` to replace the whole GeoPackage.
+(There is also a `delete_layer` parameter to overwrite an existing _layer_ with the same name.)
+
+Let's extract a selection of features from the `special_areas_conservation` layer, and add it as a second layer into the GeoPackage file:
+
+```{r}
+sac %>% 
+  filter(str_detect(sac_name, "Turnhout")) %>% # only polygons in the Turnhout area
```
Comment (ThierryO) [inline content/tutorials/spatial_standards_vector/index.Rmd]: You're filtering on attributes, not spatial location.
```suggestion
  filter(str_detect(sac_name, "Turnhout")) %>% # only polygons having 'Turnhout' in their name field
```

File: content/tutorials/spatial_standards_vector/index.Rmd [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+  select(river_name = NAAM,
+         source = BRON)
+```
+
+
+```{r paged.print=FALSE}
+habitatstreams
+```
+
+Nowadays, it is recommended to use the more recent and strict **RFC7946** implementation of GeoJSON.
```
Comment (ThierryO) [inline content/tutorials/spatial_standards_vector/index.Rmd]: How can you tell from the geoJSON file, which version is used?

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 9 above)
Comment (ThierryO) [reply content/tutorials/spatial_standards_vector/index.Rmd]: Note that diffs can be larger when rewriten the data as their order might change.

File: content/articles/geospatial_standards/index.Rmd (hunk 13 above)
[the remark this answers is not in the document]
Comment (hansvancalster) [reply content/articles/geospatial_standards/index.Rmd]: Indeed, a *personal* geodatabase is what I've used a couple of years ago and is probably closer in resemblance to the geopackage.

File: content/articles/geospatial_standards/index.Rmd [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+---
+
+Some inspiration for this post came from the beautiful books of @lovelace_geocomputation_2019, @pebesma_edzer_spatial_2019 and @heijmans_spatial_2019, and from various websites.
+
+## Why use open standards?
+
+- Open file standards ease collaboration, portability and compatibility between users, machines and applications.
+- Their (file) structure is fully documented.
+  - Consequently, scientists and programmers can build new software / packages and make innovations that use these standards, while maintaining interoperability with existing applications.
+  - And, it guarantees that your data will still be readable in a hundred years from now, independently of which IT corporation or software is dominant at that time...
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Well, at least the format is fully specified, so at least in theory I think it will always be possible to read such data. Not sure though, whether in this sense caveats still exist with binary data. You will indeed need (make/adapt) software that is capable of reading a 100-year old version of a standard, but the same holds for text data.

@peterdesmet what is your view on this? Should the statement be nuanced?

The OGC [states](https://www.opengeospatial.org/standards) this:

> These standards are the main "products" of the Open Geospatial Consortium and have been developed by the membership to address specific interoperability challenges. Ideally, when OGC standards are implemented in products or online services by two different software engineers working independently, the resulting components plug and play, that is, they work together without further debugging.

File: content/articles/geospatial_standards/index.Rmd [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+  
+Luckily, quite a list of open standards is available!
+Below, some powerful and widely-used single-file formats are introduced.
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: It has the portability advantage (one dataset = one file). It prevents from pieces getting 'lost' when sharing, republishing etc. and it is the simplest way for users to pick up / copy a dataset.

While it maximizes portability, it is not a requisite of an open standard of course. But I'm sure it will have helped these standards in becoming this popular.

File: content/articles/geospatial_standards/index.Rmd [hunk 20]
```diff
@@ -0,0 +1,90 @@
+---
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Seems logical, still I checked how well this works. The output format is the `gfm` variant of `md_document`. It appears that knitting a self-made `*.md` file in RStudio to markdown will overwrite the file with the output. This replaces bibtex references with hard-coded references and adds a hard-coded bibliography. Some other small changes occurred as well, e.g. bullet layout. IMO it is preferrable to still keep the original markdown, so Rmd seems one simple (but surely not the only) way to do this.

File: content/articles/geospatial_standards/index.Rmd [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+Though GeoJSON 2008 is obsoleted, the now recommended RFC7946 standard is still officially in a _proposal_ stage.
+That is probably the reason why GDAL does not yet default to RFC7946.
+A somehow confusing stage, it seems.
+], 
+RFC7946 is supported by the option `RFC7946=YES`.
+Here, on-the-fly reprojection to WGS84 will happen automatically.
+It applies 7 decimal places for coordinates, i.e. approximately 1 cm.
+Given the advantages, _**we advise to explicitly use RFC7946**_.
+Several functions in R allow the user to provide options that are passed to GDAL, so we can ask to deliver RFC7946 (see the [tutorial](../../tutorials/spatial_standards_vector/)).
+- In order to keep it manageable (text file size, usage in versioning systems) it can be wise to use GeoJSON for more simple cases (points and rather simple lines and polygons), and use the binary GeoPackage format for larger (more complex) cases.
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Good suggestion, will look at it.

File: content/articles/geospatial_standards/index.Rmd (hunk 20 above)
[the remark this answers is not in the document]
Comment (ThierryO) [reply content/articles/geospatial_standards/index.Rmd]: Are you rendering the md file? Because you don't need to do that.

[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Yes I rendered it. Would the bibtex references be processed on the website if I don't? Normally our index.md files already contain the processed (rendered) output, with bibtex references converted to textual references + bibliography.

File: content/articles/geospatial_standards/index.Rmd [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+], 
+so the GeoPackage is a more than suitable replacement.
+- The GeoPackage can store one or _multiple_ **vector** layers (points, lines, polygons and related feature types).
+Besides vector data, it can also store **raster** data
+or extra standalone **tables**.
+- The GeoPackage standard is [maintained](https://www.opengeospatial.org/standards/geopackage) by the [Open Geospatial Consortium](https://www.opengeospatial.org/) (OGC), which stands out as a reference when it comes to open geospatial standards.
+
+## The GeoJSON file format
+
+- One [GeoJSON](https://tools.ietf.org/html/rfc7946) file (`filename.geojson`) contains _one_ **vector** layer.
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: This must be added indeed, thanks! In case of multiple types of features it is a 'FeatureCollection'. Still, I believe this does not contradict the concept of 'one vector layer', what is stated here. One `sf` object corresponds to one 'layer', which can have multiple feature types:

``` r
st_layers("batbike_inbo.geojson")
#> Driver: GeoJSON 
#> Available layers:
#>     layer_name geometry_type features fields
#> 1 batbike_inbo                     11      7

batbike <- st_read("batbike_inbo.geojson")
#> Reading layer `batbike_inbo' from data source `/media/floris/DATA/ContinuLegen/batbike_inbo.geojson' using driver `GeoJSON'
#> Simple feature collection with 11 features and 7 fields
#> geometry type:  GEOMETRY
#> dimension:      XY
#> bbox:           xmin: 4.198113 ymin: 50.70852 xmax: 4.278424 ymax: 50.75257
#> epsg (SRID):    4326
#> proj4string:    +proj=longlat +datum=WGS84 +no_def

st_geometry_type(batbike)
#>  [1] LINESTRING POINT      POINT      POINT      POINT      POINT      POINT      POINT     
#>  [9] POINT      POINT      POINT     
#> 18 Levels: GEOMETRY POINT LINESTRING POLYGON MULTIPOINT MULTILINESTRING ... TRIANGLE
```

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 17 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: It is not hardcoded in the file which version is used, so it is not always clear. From the RFC7946 specification I understand it is a more constrained version of GeoJSON 2008, the most obvious being that WGS84 is obliged. RFC7946 does not contradict 2008, so some 2008 implementations may already conform to RFC7946 as well.

You can easily tell a GeoJSON file is _not_ RFC746 _when_ it has a `crs` key. The `crs` key is [absent](https://tools.ietf.org/html/rfc7946#appendix-B.1) from RFC7946. Other RFC7946 constraints will be less obvious to check from the file, maybe there are tools to check that.

There are also extra recommendations when applying RFC7946, but these are not obligations, so that's more difficult to rely on. A rather important one seems the [advice on coordinate precision](https://tools.ietf.org/html/rfc7946#section-11.2), which has effect in GDAL.

Hence, specifically for files generated with GDAL's default for these verions, it will be easier to distinguish both. GeoJSON 2008 by default [uses](https://gdal.org/drivers/vector/geojson.html#layer-creation-options) `COORDINATE_PRECISION=15` while it is `COORDINATE_PRECISION=7` for RFC7946.

IMHO a specific pitfall in GDAL's implementation of GeoJSON 2008 is that it will not transform data from another CRS to WGS84, while at the same time _not_ declaring `crs` in the file! (Unless when input already was WGS84, though.) Hence, when reading, WGS84 is assumed with wrong coordinates. Cf. [this post](https://github.com/r-spatial/sf/issues/344#issue-229118527).

File: content/tutorials/spatial_standards_raster/index.Rmd (hunk 15 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_raster/index.Rmd]: A matter of taste, I do like this more.

File: content/tutorials/spatial_standards_raster/index.Rmd [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+How does the result look like?
+
+```{r}
+arts
+```
+
+```{r arts}
+arts %>% 
+  as.list %>% 
+  lapply(spplot)
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/tutorials/spatial_standards_raster/index.Rmd]: Is it possible to have separate legend keys for the panels? Didn't find how to do that; is relevant for this example.

File: content/articles/geospatial_standards/index.Rmd (hunk 13 above)
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: I agree on the personal geodatabase resemblance, will mention this. Thanks!

As a sidenote, the [personal geodatabase](https://desktop.arcgis.com/en/arcmap/10.3/manage-data/geodatabases/a-comparison-of-geodatabase-types.htm) has its limitations compared to GPKG; most obvious are: windows-only, closed format, limited to 250-500 MB (GeoPackage [can](http://www.geopackage.org/spec/#_sqlite_container) handle 140 TB if the filesystem can).

Another related sidenote. Found this in the FAQ on geopackage.org:

> Q. Does GeoPackage replace Shapefile?

> A. It could but it doesn’t have to. If all you need is simple exchange and display then GeoPackage may be overkill and something like GeoJSON may be more appropriate. If you are looking for database capabilities like random access and querying then GeoPackage is a platform-independent, vendor-independent choice. GeoPackage was carefully designed this way to facilitate widespread adoption and use of a single simple file format by both commercial and open-source software applications — on enterprise production platforms as well as mobile hand-held devices.

So: they do stress the particularities of each format. In applications, books and blogs, you can observe that 'shapefile by default' is (mostly?) replaced by 'GeoPackage as default'.

File: content/articles/geospatial_standards/index.Rmd (hunk 18 above)
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: A while ago, I met a [post](https://www.linuxjournal.com/content/open-source-eternal) on this topic, which wants to confirm the general applicability of open standards still being readable far in the future:

>The only hope of reading ancient file formats is if they are completely open. That way, readers and file conversion tools can be built with relative ease. Similarly, any programs that are created for projects looking at very long-term preservation of digital material must be open source, so that they too can be examined in detail, modified and built upon. Even after just 25 years, we know that free software has won. But it is also highly likely that its success will be very long term, assuming human culture survives with any continuity. Open source—and only open source—is eternal.

File: content/articles/geospatial_standards/index.Rmd (hunk 21 above)
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: TopoJSON is an interesting topic (e.g. [this](https://github.com/r-spatial/sf/issues/381) discussion in `sf` about topology). However IMHO the implementations in R / GDAL are not mature enough:

- TopoJSON is not yet well implemented by GDAL (read-only, and it appears not to set a CRS when reading)
- a standalone implementation does exist in R by the `geojsonio` package (they cope with GeoJSON as well), and they do provide functions to read/write and work with sf-objects. However, from what I've seen, their written GeoJSON files differ from GDAL's implementation (in `geojsonio`, everything comes one line 1) and they did [not yet implement](https://github.com/ropensci/geojsonio/issues/110) RFC7946. Their generated TopoJSON file seems to suffer from the everything-on-one-line as well. I'd argue against using this package when the aim is to write files. An interesting discussion in the `sf` repo, regarding `geojsonio` is [here](https://github.com/r-spatial/sf/issues/185#issuecomment-285360564).

I prefer open standards for which a (well established) open driver is available, as this makes results comparable between applications. This is the strength of `sf`, which relies on GDAL. So for topology, we may have to wait for a well-established software (driver) solution.

TopoJSON itself has a reference implementation - a command-line JS tool by the authors - but [its specification](https://github.com/topojson/topojson-specification/blob/master/README.md) is not integrated (yet?) by open standards organizations such as OGC or IETF.

File: content/articles/geospatial_standards/index.Rmd (hunk 18 above)
[the remark this answers is not in the document]
Comment (peterdesmet) [reply content/articles/geospatial_standards/index.Rmd]: Would rephrase to:

> And, there is a much higher chance that your data will still be readable in the future.

[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Thanks for the suggestion, will write something in that sense.

File: content/articles/geospatial_standards/index.Rmd (hunk 22 above)
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Clarified by 19bb8e7

File: content/articles/geospatial_standards/index.Rmd (hunk 18 above)
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Done in de81b86

File: content/tutorials/spatial_standards_raster/index.Rmd [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+Let's create a `RasterBrick` object of three layers:
+
+```{r}
+arts <- brick(artwork) # RasterBrick with one layer (the RasterLayer from above)
+arts[[2]] <- artwork + 10 # Add second layer, e.g. based on first one
+arts[[3]] <- calc(arts[[2]], function(x) 20 ^ x) # Making third layer from second
+names(arts) <- paste0("layer", 1:3)
+```
+
+_Note: in real-life cases, you will see that the `calc()` function is more efficient than simple algebraic expressions such as for layer2._
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/tutorials/spatial_standards_raster/index.Rmd]: changed by 69e9cda

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 9 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: a0fd2e8 handles @ThierryO 's note on versioning GeoJSON

File: content/articles/geospatial_standards/index.Rmd (hunk 13 above)
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Handled by aa575f0

File: content/articles/geospatial_standards/index.Rmd (hunk 1 above)
Comment (florisvdh) [reply content/articles/geospatial_standards/index.Rmd]: Handled by 890e8a8

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 9 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: 890e8a8 handles @hansvancalster  's note on making file format differences more prominent

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 6 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: not done, see earlier remark (geojson has more complexities / limitations for the user)

File: content/tutorials/spatial_standards_vector/index.Rmd (hunk 10 above)
Comment (florisvdh) [reply content/tutorials/spatial_standards_vector/index.Rmd]: Handled by 890e8a8

**Comment by florisvdh** [comment]
[Herewith](https://github.com/inbo/tutorials/files/4149052/geospatial_standards.zip) newly compiled html output.

**Comment by peterdesmet** [comment]
Is this online yet? Wanna read it. 😄

**Comment by florisvdh** [comment]
Just :wink:
